### DISCUSSION TITLE
register as a function, vs. called without arguments

### DISCUSSION THREAD

**Discussion by dandv** [react-hook-form/react-hook-form#1684] [General] [answered]
I've noticed that in the [nested fields example](https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-x7btr?file=/src/nestedFieldArray.js), `register()` needs to be called without arguments. In most other examples, I see it called either with specific arguments, or just left as a function parameter:

   <input name="example" defaultValue="test" ref={register} />

If I remove the `()` from the nested fields example, I get this error if I append 2 nested fields then delete them both:

> data.slice is not a function or its return value is not iterable

![image](https://user-images.githubusercontent.com/33569/82742762-e1b81380-9d16-11ea-8960-94d11f825587.png)

Just wanted to understand this `register` vs `register()` situation. Thanks!

**Comment by bluebill1049** [comment, marked as answer]
This problem haunting us for a long time. Going `register()` is our last option on the table. 

`Ref` gets invoked after input finished mount, and invoked again when it gets removed. However, during `.map` React reconcile kicks in and optimize when to remove/append a real Node, so when input shuffle and swap we want the correct input to get `re-register` again. `register()` will force that to call as it's a new anonymous `callback`.

**Reply by dandv** [reply]
It looks like I'm still doing something wrong, because I get that `data.slice is not a function` error [even when I call `register()`](https://codesandbox.io/s/react-hook-form-usefieldarray-error-on-removal-hc8p2?file=/src/fieldArray.js).

**Reply by bluebill1049** [reply]
your codesandbox is broken.

**Reply by bluebill1049** [reply]
have a play with this cdoesandbox: https://codesandbox.io/s/react-hook-form-usefieldarray-vy8fv

**Reply by dandv** [reply]
Sorry, I've just fixed the link, here's the smallest [CSB](https://codesandbox.io/s/react-hook-form-usefieldarray-error-on-removal-hc8p2?file=/src/fieldArray.js) that reproduces the error.

**Reply by bluebill1049** [reply]
either pass `regsiter` function down as prop or using `React.forwardRef`

https://reactjs.org/docs/forwarding-refs.html

https://codesandbox.io/s/react-hook-form-usefieldarray-error-on-removal-ci33u?file=/src/fieldArray.js

**Reply by dandv** [reply]
I see, thanks, but the last CSB still fails with that error if you add 2 fields then delete both.

**Reply by bluebill1049** [reply]
i missed the `register()`

https://codesandbox.io/s/react-hook-form-usefieldarray-error-on-removal-ci33u?file=/src/DeletableField.js

**Reply by dandv** [reply]
That worked with v5, but after upgrading to v6, fails with the same error.

https://codesandbox.io/s/react-hook-form-usefieldarray-error-on-removal-rqnml

**Reply by bluebill1049** [reply]
what's the steps to reproduce?

**Reply by dandv** [reply]
Turns out error was on my side (sorry), but in [your code](https://codesandbox.io/s/react-hook-form-usefieldarray-error-on-removal-ci33u?file=/src/DeletableField.js), without changing anything, I notice `onSubmit` appears to be called every time I click on `ADD`, in addition to when I click SUBMIT. Not very important, just FYI.

**Reply by bluebill1049** [reply]
i think one of the buttons, i must forget to include `type="button"`
